Fix metrics behavior when either true_locs or est_locs is empty

`match_by_locs()` throws a runtime error if either true_locs or est_locs are empty, which can happen if the priors are too small. If both are empty we can just continue in `update()`, since there are no TP/FP/TN/FN. If one is empty and the other isn't, we probably need to update the values manually.

```
# Penalize all pairs which are greater than slack apart to favor valid matches.
>       locs_err = locs_err + (locs_err_l_infty > slack) * locs_err.max()
E       RuntimeError: max(): Expected reduction dim to be specified for input.numel() == 0. Specify the reduction dim with the 'dim' argument.
```

--- a/bliss/metrics.py
+++ b/bliss/metrics.py
@@ -72,6 +72,17 @@ class BlissMetrics(Metric):
             ntrue, nest = true.n_sources[b].int().item(), est.n_sources[b].int().item()
             tlocs, elocs = true.plocs[b], est.plocs[b]
             tgbool, egbool = true["galaxy_bools"][b].reshape(-1), est["galaxy_bools"][b].reshape(-1)
+
+            self.total_true_n_sources += ntrue  # type: ignore
+            # if either ntrue or nest are 0, manually increment FP/FN and continue
+            if ntrue == 0 or nest == 0:
+                if nest > 0:  # all estimated are false positives
+                    self.detection_fp += nest
+                    self.gal_fp += egbool.sum()
+                elif ntrue > 0:  # all true are false negatives
+                    self.gal_fn += tgbool.sum()
+                continue
+
             mtrue, mest, dkeep, avg_distance = match_by_locs(
                 tlocs[: int(ntrue)], elocs[: int(nest)], self.slack
             )
@@ -83,7 +94,6 @@ class BlissMetrics(Metric):
             self.detection_tp += detection_tp
             self.detection_fp += detection_fp
             self.avg_distance += avg_distance
-            self.total_true_n_sources += ntrue  # type: ignore
             conf_matrix = confusion_matrix(tgbool.cpu(), egbool.cpu(), labels=[1, 0])
             # decompose confusion matrix to pass to lightning logger
             self.gal_tp += conf_matrix[0][0]

--- a/tests/test_metrics.py
+++ b/tests/test_metrics.py
@@ -163,6 +163,32 @@ class TestMetrics:
         assert class_acc == 1 / 2
         assert avg_distance.item() == 50 * (0.01 + (0.01 + 0.09) / 2) / 2
 
+    def test_no_sources(self):
+        """Tests that metrics work when there are no true or estimated sources."""
+        metrics = BlissMetrics(slack=2.0)
+
+        true_locs = torch.tensor(
+            [[[10, 10]], [[20, 20]], [[30, 30]], [[40, 40]]], dtype=torch.float
+        )
+        est_locs = torch.tensor([[[10, 10]], [[20, 20]], [[30, 30]], [[41, 41]]], dtype=torch.float)
+        true_galaxy_bools = torch.tensor([[[1]], [[1]], [[1]], [[1]]])
+        est_galaxy_bools = torch.tensor([[[1]], [[1]], [[1]], [[1]]])
+        true_sources = torch.tensor([0, 0, 1, 1])
+        est_sources = torch.tensor([0, 1, 0, 1])
+
+        d_true = {"n_sources": true_sources, "plocs": true_locs, "galaxy_bools": true_galaxy_bools}
+        true_params = FullCatalog(50, 50, d_true)
+
+        d_est = {"n_sources": est_sources, "plocs": est_locs, "galaxy_bools": est_galaxy_bools}
+        est_params = FullCatalog(50, 50, d_est)
+
+        results = metrics(true_params, est_params)
+
+        assert results["detection_precision"] == 1 / 2
+        assert results["detection_recall"] == 1 / 2
+        assert results["gal_tp"] == results["gal_fp"] == results["gal_fn"] == 1
+        assert results["avg_distance"].item() == 1
+
     def test_photo_self_agreement(self, catalogs):
         """Compares PhotoFullCatalog to itself as safety check for metrics."""
         results = BlissMetrics()(catalogs["photo"], catalogs["photo"])

